src/pudl/transform/phmsagas.py

```diff
@@ -0,0 +1,702 @@
+        df[state_col] = (
+            df[state_col]
+            .str.strip()
+            .replace(state_to_abbr)
+            .where(df[state_col].isin(state_to_abbr.values()), pd.NA)
+        )
+
+    # Standardize zip codes
+    df["office_zip"] = zero_pad_numeric_string(df["office_zip"], n_digits=5)
+    df["headquarters_zip"] = zero_pad_numeric_string(df["headquarters_zip"], n_digits=5)
+
+    # Drop duplicates
+    df = df.drop_duplicates()
+
+    # Identify non-unique groups based on our PK
+    # There are a small number of records with duplicate report numbers
+    # and either duplicate or "test" data in the rows. Remove the 'bad data'
+    # to ensure a unique primary key.
+    non_unique_groups = df[df.duplicated(["report_number"], keep=False)]
+
+    # Apply some custom filtering logic to non-unique groups
+    filtered_non_unique_rows = non_unique_groups.groupby(
+        ["report_number"], group_keys=False
+    ).apply(combined_filter)
+
+    # There are ten values that we expect to be duplicated.
+    assert len(filtered_non_unique_rows) <= 113, (
+        f"Found {len(filtered_non_unique_rows)} records with duplicates, expected 113 or less."
+    )
+
+    # Combine filtered non-unique rows with untouched unique rows
+    unique_rows = df.drop(non_unique_groups.index)
+    df = pd.concat([unique_rows, filtered_non_unique_rows], ignore_index=True)
+
+    return df
+
+
+def filter_if_test_in_address(group: pd.DataFrame) -> pd.DataFrame:
+    """Filters out rows with "test" in address columns.
+
+    For any group of rows with the same combination of "operator_id_phmsa"
+    and "report_number", if at least one row in the group does not contain the string
+    "test" (case-insensitive) in either "office_street_address" or
+    "headquarters_street_address", keep only the rows in the group that do not contain
+    "test" in these columns. If all rows in the group contain "test" in either of the
+    columns, leave the group unchanged.
+
+    Args:
+        group: A grouped subset of the DataFrame.
+
+    Returns:
+        The filtered group of rows.
+    """
+    # Check if at least one row in the group does NOT contain "test" in both of the specified columns
+    contains_test = group.apply(
+        lambda row: "test" in str(row["office_street_address"]).lower()
+        or "test" in str(row["headquarters_street_address"]).lower(),
+        axis=1,
+    )
+    has_non_test = not contains_test.all()
+
+    if has_non_test:
+        # Keep rows where "test" does NOT appear in either column
+        return group[~contains_test]
+    # If all rows have "test", keep the group as is
+    return group
+
+
+def filter_by_city_in_name(group: pd.DataFrame) -> pd.DataFrame:
+    """Deduplication to keep duplicated rows where city and operator name overlap.
+
+    Filter to only keep rows where "office_city" value is contained in the
+    "operator_name_phmsa" value (case insensitive).
```

Comment:
the changed i made in this function imho shouldn't stand.... I'd like to fully re-think the de-duplication of the pk. i had to do some back flipping to make this work because the older years of data have many more nulls and many more duplicates on the `report_number`... 

```python
non_unique_groups.report_year.value_counts().sort_index()

report_year
1979.0     8
1980.0    86
1981.0    44
1982.0     2
Name: count, dtype: Int641
```

[non_unique_groups.pkl.zip](https://github.com/user-attachments/files/22467683/non_unique_groups.pkl.zip)